src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CompressedMessageBinaryEncoder.cs

```diff
@@ -0,0 +1,226 @@
+            _encoderSettings = encoderSettings; // can be null
+        }
+
+        /// <summary>
+        /// Gets a flag whether the message compressed or not.
+        /// </summary>
+        /// <returns>A compressed flag.</returns>
+        public bool IsCompressed()
+        {
+            var stream = CreateBinaryReader().BsonStream;
+
+            var position = stream.Position;
+
+            // Skip header to opcode
+            stream.ReadInt32();
+            stream.ReadInt32();
+            stream.ReadInt32();
+            var isCompressed = stream.ReadInt32() == (int)Opcode.Compressed;
+
+            stream.Position = position;
+
+            return isCompressed;
+        }
+
+        /// <summary>
+        /// Reads the message.
+        /// </summary>
+        /// <returns>A message.</returns>
+        public CompressedMessage ReadMessage()
+        {
+            var stream = CreateBinaryReader().BsonStream;
+            var startPosition = stream.Position;
+            var compressedMessageInfo = ReadMessageHeaders(stream);
+
+            using (var chunkSource = new OutputBufferChunkSource(BsonChunkPool.Default))
+            {
+                var buffer = new MultiChunkBuffer(chunkSource);
+                var originalMessageStream = new ByteBufferStream(buffer);
+
+                originalMessageStream.WriteInt32(compressedMessageInfo.OriginalMessageSize + MessageHeaderLength); // original message size with header
+                originalMessageStream.WriteInt32(compressedMessageInfo.RequestId);
+                originalMessageStream.WriteInt32(compressedMessageInfo.ResponseTo);
+                originalMessageStream.WriteInt32((int)compressedMessageInfo.OriginalOpCode);
+
+                var compressor = _compressorSource.Get(compressedMessageInfo.CompressorType);
+                // some compressors (`snappy`) can fail here for plural input messages,
+                // but since we don't read plural compressed messages in the current implementation,
+                // we can ignore it
+                compressor.Decompress(stream, originalMessageStream);
+                // set the end position for stream
+                stream.Position = startPosition + compressedMessageInfo.CompressedMessageLength;
+
+                originalMessageStream.Position = 0;
+                buffer.MakeReadOnly();
+                var encoder = GetEncoderByOpcode(compressedMessageInfo.OriginalOpCode, originalMessageStream);
+                var originalMessage = encoder.ReadMessage();
+
+                return new CompressedMessage(
+                    originalMessage,
+                    originalMessageStream,
+                    compressedMessageInfo.CompressorType);
+            }
+        }
+
+        /// <summary>
+        /// Writes the message.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        public void WriteMessage(CompressedMessage message)
```

Comment:
I propose this implementation instead:

        public void WriteMessage(CompressedMessage message)
        {
            Ensure.IsNotNull(message, nameof(message));

            var writer = CreateBinaryWriter();
            var stream = writer.BsonStream;

            var uncompressedMessageStream = message.OriginalMessageStream;
            var uncompressedMessageLength = uncompressedMessageStream.ReadInt32();
            var requestId = uncompressedMessageStream.ReadInt32();
            var responseTo = uncompressedMessageStream.ReadInt32();
            var originalOpcode = (Opcode)uncompressedMessageStream.ReadInt32();

            var compressorType = message.CompressorType;
            var compressor = _compressorSource.Get(compressorType);

            var messageStartPosition = stream.Position;
            stream.WriteInt32(0); // messageLength
            stream.WriteInt32(requestId);
            stream.WriteInt32(responseTo);
            stream.WriteInt32((int)Opcode.Compressed);
            stream.WriteInt32((int)originalOpcode);
            stream.WriteInt32(uncompressedMessageLength - MessageHeaderLength);
            stream.WriteByte((byte)compressorType);
            compressor.Compress(uncompressedMessageStream, stream);
            stream.BackpatchSize(messageStartPosition);
        }

Reply:
done